### PR TITLE
Refactor Debian Package Build logic

### PR DESCRIPTION
**Pull request by msheiny**
## Status

🚥 Ready for Review 🚥 

## Description of Changes

So I first started this PR to address #1825 BUTT in order to do that I needed to start converting the CI logic to moleculev2 (believe me it makes things much easier ). This branch started getting too big so I figured lets start testing it as-is and pull a PR. *ahem* 

Changes proposed in this pull request:

So what does this PR contain? It converts the current docker build logic for creating debian packages:

- Utilizes a non-privileged Ubuntu trusty docker image (vs. previously utilized Debian jessie image)
- Provides tooling for local dev and removes need for Vagrant build flow
- Takes advantage of molecule which will provide much easier to maintain in the long-run.

## Testing

In order to test the following things need to occur:

- Deb packages need to get thrown against CI ✔️ 
- a full vagrant run of the staging environment locally needs to occur

## Deployment

Doesnt really affect deployments in the wild BUT this does affect building debian packages for deployment (by SD maintainers)

### If you made changes to documentation:

- [x] Doc linting passed locally

### CODE REVIEW

**Comment by dachary** [comment]
I was confused by this PR until I realized molecule is used both for testing and to produce debian packages. I think it would be saner to have a role whose sole purpose is to build deb packages using docker. And molecule testing this role actually produces the artifcats it is supposed to. Does that make sense ?

**Comment by msheiny** [comment]
> I was confused by this PR until I realized molecule is used both for testing and to produce debian packages. I think it would be saner to have a role whose sole purpose is to build deb packages using docker. And molecule testing this role actually produces the artifcats it is supposed to. Does that make sense ?

Well I mean there is already a playbook `install_files/ansible-base/build-deb-pkgs.yml` and several ansible roles that do that:
* ./install_files/ansible-base/roles/build-securedrop-app-code-deb-pkg
* ./install_files/ansible-base/roles/build-generic-pkg
* ./install_files/ansible-base/roles/build-ossec-deb-pkg

`molecule/builder/playbook.yml` is a symlink to `install_files/ansible-base/build-deb-pkgs.yml`

This molecule scenario just takes advantage of all of those playbooks/roles + provides a scaffolding for running the testinfra tests afterwards. You just have to run `make build-debs` now vs. `vagrant up build`.   If you really want to skip testing and just build the debs you can run `molecule converge -s builder`

**Comment by dachary** [comment]
I can't help but think using a test framework as a tool to produce artifacts (deb in this case) is a clever hack that we'll eventually regret. Why not just make it so `make build-debs` provisions docker instead of vagrant + builds the packages without requiring molecule ? And use molecule solely for the purpose of testing that.

**Comment by msheiny** [comment]
The nice thing about moleculev2 is that there is very little abstraction being done here -- and I will concede that one of the things I didnt like about moleculev1 was that there was a ton of weird abstraction going on. In fact, I remember @conorsch was arguing when I laid out the CI scripts that are currently being used that we should use molecule v1 -- but it was just a huge pain in the butt.

BUT Moleculev2 in itself is basically a clever wrapper around a bunch of other tools: vagrant, docker, ansible, testinfra, flake8, ansible-lint. I was essentially doing what molecule is doing before with bash scripts. The problem with controlling this logic with bash scripts is that we have to do this same thing across multiple repos. So overtime  I start to get a little better at organizing the scripts with every new repo, and when I look back at older repos it doesnt make sense. So then I'm maintaining and updating a slew of bash scripts across multiple repos -- which is super painful and leads to confusion when people go to update. Using something like molecule brings a standard to the chaos -- and is very easy to rip out later if we deem it a problem.

Basically I am doing what molecule is doing but using my own shell

**Comment by msheiny** [comment]
Tested locally with staging VMs 👍

**Comment by dachary** [comment]
> The problem with controlling this logic with bash scripts is that we have to do this same thing across multiple repos.

I was not suggesting to revert to bash scripts :-)

> is very easy to rip out later if we deem it a problem.

I respectfully disagree and encourage you to consider an alternative approach.

**Comment by conorsch** [comment]
This PR doesn't close an issue, which makes it difficult to review. That seems to be part of the problem with understanding its intent, and how effectively it meets the implicit goals. 

> start converting the CI logic to moleculev2 (believe me it makes things much easier )

@msheiny We need a discrete issue for the Molecule v2 migration. I'll open one. I agree with you that migrating to Molecule v2 would allow us to delete a ton of custom bash from the repo, but there's no issue tracking that discussion, so it's difficult for other folks to follow along. 

Also let's consider again the fact that we have a dozen or so _custom_ roles in this repository, which means we're not making use of community-maintained or even FPF-maintained configuration logic, which makes refactors like this one cumbersome. Now that we have stronger configuration testing, I would be happy ripping out the custom roles in favor of reusable ones (likely via `git subtree`) and rely on the config tests to guard against regressions.

> I was not suggesting to revert to bash scripts :-)

@dachary Unfortunately that's what we've got right now, for the CI tooling located in the `devops/` directory. Moving to Molecule v2 is a great idea for CI, but again, that discussion belongs in a separate issue. 

> I can't help but think using a test framework as a tool to produce artifacts (deb in this case) is a clever hack that we'll eventually regret.

A fair point! However we're already using Vagrant, a tool intended for developer environments, to build debs, and have been doing so for years. By swapping in Molecule for controlling the Vagrant logic, we get much more powerful testing and linting for free. Still interested in adding stronger validation of the build process (e.g. #308, #1637), and adding those features _without_ including Molecule would involve a ton of custom scripting, and I'd much rather rely on a strong community-maintained solution.

**Comment by msheiny** [comment]
rebased off `develop` to solve potential merge conflict

**Comment by msheiny** [comment]
> This PR doesn't close an issue, which makes it difficult to review. That seems to be part of the problem with understanding its intent, and how effectively it meets the implicit goals.

👍 -- @conorsch see #2198

**Comment by msheiny** [comment]
@conorsch ping

**Comment by conorsch** [comment]
Reviewing!

**Comment by conorsch** [comment]
@msheiny Could you rebase this on top of latest develop? I'm already seeing some logic missing from this branch, e.g. #2191.

**Comment by msheiny** [comment]
rebased 👍

**Comment by conorsch** [comment]
@heartsucker Would appreciate a functional test from you, meaning confirming that staging still works on your machine. You should: 

1. Destroy all local VMs.
2. Check out this feature branch.
3. Run `make build-debs`. 
4. Run `vagrant provision /staging/`.
5. Run config tests for app-staging and mon-staging.
6. Run app tests on app-staging (including `--page-layout` tests). 

If any errors show up, please post them here.

File: install_files/ansible-base/securedrop-staging.yml [hunk 1]
```diff
@@ -1,10 +1,22 @@
...
+
+    - set_fact:
+        expected_debs: "{{hostvars['app-staging']['local_deb_packages'] + hostvars['mon-staging']['local_deb_packages']}}"
+        found_debs: "{{ search_deb_results.files | map(attribute='path') | list }}"
+
+    - name: Fail if package is missing from build dir
+      fail:
+        msg: "run 'make build-debs' - ../../build/{{item}} missing"
+      when: "'../../build/'+item not in found_debs"
+      with_items: "{{ expected_debs }}"
```
Comment (conorsch) [inline install_files/ansible-base/securedrop-staging.yml]: Nit: It would be slightly more readable to use a `|basename` filter on the `found_debs` vars declaration, then the `when` conditional wouldn't need to prefix the build path.

File: molecule/builder/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,7 @@
+FROM ubuntu:14.04
+
+RUN apt-get -y update && apt-get install -y python rsync sudo bash \
+            devscripts git aptitude ipython libssl-dev ntp \
+            python-dev python-pip python-ipdb ruby tmux vim paxctl && \
+            aptitude -y dist-upgrade && \
+            apt-get clean
```
Comment (conorsch) [inline molecule/builder/Dockerfile]: Bummer to have these hard-coded rather than vars-based, but that's Docker for ya. Worth adding some config tests to track these explicitly, since many of them were present implicitly in the Ubuntu Trusty build VM.

File: molecule/builder/create.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    docker_hash: "{{ lookup('file', '../docker_hashes.yml') | from_yaml }}"
+  tasks:
+    - name: Create molecule instance(s)
+      docker_container:
+        name: "{{ item.name }}"
+        hostname: "{{ item.name }}"
+        image: "{{ item.image }}@{{ docker_hash[item.hash_name] }}"
+        state: started
+        recreate: True
+        command: "tail -f /dev/null"
```
Comment (conorsch) [inline molecule/builder/create.yml]: Crafty. I would have used `sleep infinity` instead, but either works fine for a forever process.

File: molecule/builder/tests/conftest.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import yaml
+
+
+def pytest_namespace():
+    """ Return dict of vars imported as 'securedrop_test_vars' into pytest
+        global namespace
+    """
+    filepath = os.path.join(os.path.dirname(__file__), "vars.yml")
+    with open(filepath, 'r') as f:
+        return dict(securedrop_test_vars=yaml.safe_load(f))
```
Comment (conorsch) [inline molecule/builder/tests/conftest.py]: Hmm, we'll need to duplicate these conftest.py files across multiple scenarios in Molecule (has bearing on #2218). I wonder if there's a better way?

Comment (msheiny) [reply molecule/builder/tests/conftest.py]: Well FYI , #2218 doesnt need this because I was unable to use molecule to kick off the testinfra tests there in a way that wouldn't cause a huge refactor. In that PR, we are calling the testinfra tests with a similar shell script that we currently have (took existing and stripped away unneeded bits)

File: docs/development/getting_started.rst [hunk 5]
```diff
@@ -128,6 +137,7 @@ different version, the path to ``virtualenvwrapper.sh`` will differ. Running
     source /usr/local/bin/virtualenvwrapper.sh
     mkvirtualenv -p python2 securedrop
     pip install -r securedrop/requirements/develop-requirements.txt
+    pip install -r testinfra/requirements.txt
```
Comment (conorsch) [inline docs/development/getting_started.rst]: While testing this PR branch, I intuitively reran the `develop-requirements` files and confirmed that Molecule was _not_ added, even though it's a requirement for development now. We should fix the requirements story quite soon. Willing to let it pass here and in #2218, and continue tracking in #1918 & #2175.

**Comment by dachary** [comment]
> Run make build-debs.

It requires permissions to talk to the docker daemon which is not the case for the non-root users. It can be fixed easily by adding the user to the docker group and then running newgrp docker. It would be nice to have a more explicit message about that. Currently it looks like
<pre>
    TASK [Destroy molecule instance(s)] ********************************************
    failed: [localhost] (item=(censored due to no_log)) => {"censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result"}
</pre>
and after disabling the
<pre>
#  no_log: "{{ not lookup('env', 'MOLECULE_DEBUG') | bool }}"
</pre>
line in **molecule/builder/destroy.yml** it looks like
<pre>
    TASK [Destroy molecule instance(s)] ********************************************
failed: [localhost] (item={'image': u'quay.io/freedomofpress/ubuntu:trusty', 'hash_name': u'ubuntu', 'name': u'build', 'groups': [u'build']}) => {"failed": true, "item": {"groups": ["build"], "hash_name": "ubuntu", "image": "quay.io/freedomofpress/ubuntu:trusty", "name": "build"}, "msg": "Error connecting: Error while fetching server API version: ('Connection aborted.', error(13, 'Permission denied'))"}
</pre>
which is not really more explicit.

**Comment by heartsucker** [comment]
@conorsch Am testing now.

**Comment by heartsucker** [comment]
Other than this one error, everything works.

```
(venv) heartsucker@pythagoras:~/code/freedom-of-press/securedrop (StabilizeCI)$ ./testinfra/test.py app-staging
Running Testinfra suite against 'app-staging'...
Target roles:
    - testinfra/app
    - testinfra/app-code
    - testinfra/common
    - testinfra/development/test_xvfb.py
Traceback (most recent call last):
  File "/home/heartsucker/code/freedom-of-press/securedrop/venv/local/lib/python2.7/site-packages/_pytest/config.py", line 379, in _importconftest
    mod = conftestpath.pyimport()
  File "/home/heartsucker/code/freedom-of-press/securedrop/venv/local/lib/python2.7/site-packages/py/_path/local.py", line 680, in pyimport
    raise self.ImportMismatchError(modname, modfile, self)
ImportMismatchError: ('conftest', '/vagrant/testinfra/conftest.py', local('/home/heartsucker/code/freedom-of-press/securedrop/testinfra/conftest.py'))
ERROR: could not load /home/heartsucker/code/freedom-of-press/securedrop/testinfra/conftest.py

Traceback (most recent call last):
  File "./testinfra/test.py", line 136, in <module>
    run_testinfra(target_host)
  File "./testinfra/test.py", line 132, in run_testinfra
    subprocess.check_call(testinfra_command)
  File "/usr/lib/python2.7/subprocess.py", line 186, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['testinfra', '-vv', '-n', 'auto', '--connection', 'ansible', '--ansible-inventory', '.vagrant/provisioners/ansible/inventory/vagrant_ansible_inventory', '--hosts', 'app-staging', 'testinfra/app', 'testinfra/app-code', 'testinfra/common', 'testinfra/development/test_xvfb.py']' returned non-zero exit status 4
```

**Comment by conorsch** [comment]
@heartsucker Run the command `./testinfra/test.py app-staging` from your host machine, not the VM guest. Then it should work for you! 

The failed attempt to run the testinfra checks inside the guest may have generated conflicting cache files—try running `find -type f -iname '*.pyc' -delete'` from your host machine prior to rerunning testinfra.

**Comment by conorsch** [comment]
@heartsucker My apologies, I misread your terminal session. Looks like you did indeed run the command from your host machine, not from the guest. Try the pyc purge and let me know if that helps.

**Comment by heartsucker** [comment]
Yes, that `find` command did it. How annoying.

Anywho. Green across the board.

**Comment by redshiftzero** [comment]
Note that a testinfra test for the development environment needs updating here

**Comment by conorsch** [comment]
> Note that a testinfra test for the development environment needs updating here

Hmm, all tests were passing previously, must have been the rebase. Go ahead and fix, I'm happy to re-review.

**Comment by redshiftzero** [comment]
Got passing tests on `11e8e0e` locally. Let's see if Travis agrees.

**Comment by dachary** [comment]
I confirm the following instructions (just modified to make them a little more explicit) worked for me as well :-)

Make sure docker ps works (without sudo)

1. Destroy all local VMs.
2. Run `pip install -r testinfra/requirements.txt`
2. Check out this feature branch.
3. Run `make build-debs`. 
4. Run `vagrant provision /staging/`.
5. Run config tests for app-staging and mon-staging testinfra/test.py app-staging ; testinfra/test.py mon-staging
6. Run app tests on app-staging (including `--page-layout` tests). 
<pre>
vagrant@app-staging:~$ sudo bash
root@app-staging:~# cd /var/www/securedrop/
root@app-staging:/var/www/securedrop# pytest -v --page-layout tests
</pre>
